File: app/src/main/java/com/example/tradeup/ui/auth/LoginFragment.java
package com.example.tradeup.ui.auth;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.fragment.NavHostFragment;

import com.example.tradeup.R;
import com.example.tradeup.core.utils.Event;
import com.example.tradeup.core.utils.FcmTokenUtil;
import com.example.tradeup.databinding.FragmentLoginBinding;
import com.example.tradeup.services.MyFirebaseMessagingService;
import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.messaging.FirebaseMessaging;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class LoginFragment extends Fragment {

    private static final String TAG = "LoginFragment";

    private FragmentLoginBinding binding;
    private AuthViewModel viewModel;

    private ActivityResultLauncher<Intent> googleSignInLauncher;
    private GoogleSignInClient googleSignInClient;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        viewModel = new ViewModelProvider(this).get(AuthViewModel.class);

        configureGoogleSignIn();
        setupGoogleSignInLauncher();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentLoginBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setupListeners();
        setupUI();
        checkInitialInputState();
        setupObservers();
    }

    private void setupUI() {
        String rememberedEmail = viewModel.getRememberedEmail();
        if (rememberedEmail != null) {
            binding.editTextEmailLogin.setText(rememberedEmail);
            binding.checkboxRememberMe.setChecked(true);
        }
    }

    private void setupListeners() {
        binding.buttonLogin.setOnClickListener(v -> {
            String email = binding.editTextEmailLogin.getText().toString().trim();
            String password = binding.editTextPasswordLogin.getText().toString().trim();
            boolean rememberMe = binding.checkboxRememberMe.isChecked();
            viewModel.loginUser(email, password, rememberMe);
        });

        binding.buttonGoogleLogin.setOnClickListener(v -> {
            Intent signInIntent = googleSignInClient.getSignInIntent();
            googleSignInLauncher.launch(signInIntent);
        });

        binding.textViewForgotPassword.setOnClickListener(v ->
                NavHostFragment.findNavController(this).navigate(R.id.action_loginFragment_to_forgotPasswordFragment)
        );

        binding.layoutRegisterLink.setOnClickListener(v ->
                NavHostFragment.findNavController(this).navigate(R.id.action_loginFragment_to_registerFragment)
        );

        TextWatcher loginTextWatcher = new TextWatcher() {
            @Override public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
            @Override public void afterTextChanged(Editable s) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                String emailInput = binding.editTextEmailLogin.getText().toString().trim();
                String passwordInput = binding.editTextPasswordLogin.getText().toString().trim();
                binding.buttonLogin.setEnabled(!emailInput.isEmpty() && !passwordInput.isEmpty());
            }
        };

        binding.editTextEmailLogin.addTextChangedListener(loginTextWatcher);
        binding.editTextPasswordLogin.addTextChangedListener(loginTextWatcher);
    }

    private void checkInputState() {
        String emailInput = binding.editTextEmailLogin.getText().toString().trim();
        String passwordInput = binding.editTextPasswordLogin.getText().toString().trim();
        binding.buttonLogin.setEnabled(!emailInput.isEmpty() && !passwordInput.isEmpty());
    }

    private void checkInitialInputState() {
        binding.buttonLogin.setEnabled(false);
        checkInputState();
    }

    private void setupObservers() {
        viewModel.isLoading().observe(getViewLifecycleOwner(), isLoading -> {
            binding.progressBarLogin.setVisibility(isLoading ? View.VISIBLE : View.GONE);
            binding.buttonLogin.setEnabled(!isLoading);
            binding.buttonGoogleLogin.setEnabled(!isLoading);
        });

        viewModel.getToastMessage().observe(getViewLifecycleOwner(), event -> {
            String message = event.getContentIfNotHandled();
            if (message != null) {
                Toast.makeText(getContext(), message, Toast.LENGTH_LONG).show();
            }
        });

        viewModel.getNavigationEvent().observe(getViewLifecycleOwner(), event -> {
            if (event.getContentIfNotHandled() != null) {
                // Sau khi đăng nhập thành công, lấy và lưu token FCM
                getAndSaveFcmToken();
                NavHostFragment.findNavController(this).navigate(R.id.action_global_to_main_nav);
            }
        });
    }

    private void configureGoogleSignIn() {
        GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(getString(R.string.default_web_client_id))
                .requestEmail()
                .build();
        googleSignInClient = GoogleSignIn.getClient(requireActivity(), gso);
    }

    private void setupGoogleSignInLauncher() {
        googleSignInLauncher = registerForActivityResult(
                new ActivityResultContracts.StartActivityForResult(),
                result -> {
                    if (result.getResultCode() == Activity.RESULT_OK) {
                        Task<GoogleSignInAccount> task = GoogleSignIn.getSignedInAccountFromIntent(result.getData());
                        handleGoogleSignInResult(task);
                    } else {
                        Log.w(TAG, "Google Sign-In cancelled by user.");
                    }
                });
    }

    private void handleGoogleSignInResult(Task<GoogleSignInAccount> completedTask) {
        try {
            GoogleSignInAccount account = completedTask.getResult(ApiException.class);
            if (account != null && account.getIdToken() != null) {
                Log.d(TAG, "Google Sign-In successful, got ID Token.");
                viewModel.loginWithGoogle(account.getIdToken());
            } else {
                String errorMessage = "Google Sign-In failed: ID Token is null.";
                Log.e(TAG, errorMessage);
                Toast.makeText(getContext(), errorMessage, Toast.LENGTH_SHORT).show();
            }
        } catch (ApiException e) {
            String errorMessage = "Google Sign-In failed with code: " + e.getStatusCode();
            Log.e(TAG, errorMessage, e);
            Toast.makeText(getContext(), "Google Sign-In failed. Please try again.", Toast.LENGTH_SHORT).show();
        }
    }

    // Phương thức mới để lấy và lưu token FCM
    private void getAndSaveFcmToken() {
        FirebaseMessaging.getInstance().getToken().addOnCompleteListener(task -> {
            if (task.isSuccessful()) {
                String token = task.getResult();
                Log.d(TAG, "FCM token retrieved: " + token);
                String userId = FirebaseAuth.getInstance().getCurrentUser().getUid();
                FcmTokenUtil.saveFcmTokenToFirestore(userId, token);
            } else {
                Log.w(TAG, "Failed to retrieve FCM token", task.getException());
                if (getContext() != null) {
                    Toast.makeText(getContext(), "Failed to retrieve FCM token.", Toast.LENGTH_SHORT).show();
                }
            }
        });
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: app/src/main/java/com/example/tradeup/ui/auth/LoginFragment.kt
package com.example.tradeup.ui.auth // Đảm bảo package này đúng

// QUAN TRỌNG: THAY THẾ ĐƯỜNG DẪN NÀY BẰNG ĐƯỜNG DẪN ĐÚNG ĐẾN FILE SessionManager.kt CỦA BẠN
import android.os.Bundle
import android.util.Log
import android.util.Patterns
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.core.view.isVisible
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.findNavController
import com.example.tradeup.R
import com.example.tradeup.core.utils.SessionManager
import com.example.tradeup.databinding.FragmentLoginBinding
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject

@AndroidEntryPoint
class LoginFragment : Fragment() {

    private var _binding: FragmentLoginBinding? = null
    private val binding get() = _binding!!

    private val authViewModel: AuthViewModel by viewModels()

    @Inject // Hilt sẽ inject SessionManager được cung cấp bởi AppModule
    lateinit var sessionManager: SessionManager

    companion object {
        private const val TAG = "LoginFragment"
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentLoginBinding.inflate(inflater, container, false)
        Log.d(TAG, "onCreateView called")
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        Log.d(TAG, "onViewCreated called")

        loadRememberedCredentials()
        setupClickListeners()
        setupObservers()
    }

    private fun loadRememberedCredentials() {
        // Kiểm tra xem sessionManager đã được inject chưa (để tránh NullPointerException nếu có vấn đề với Hilt)
        if (::sessionManager.isInitialized) {
            if (sessionManager.shouldRememberMe()) {
                binding.editTextEmailLogin.setText(sessionManager.getEmail())
                binding.checkboxRememberMe.isChecked = true
                Log.d(TAG, "Loaded remembered email: ${sessionManager.getEmail()}")
            } else {
                binding.checkboxRememberMe.isChecked = false
                Log.d(TAG, "Remember me was false.")
            }
        } else {
            Log.e(TAG, "SessionManager not initialized in loadRememberedCredentials. Hilt injection might have failed.")
        }
    }

    private fun setupClickListeners() {
        Log.d(TAG, "setupClickListeners called")

        binding.buttonLogin.setOnClickListener {
            Log.d(TAG, "Login button clicked")
            val email = binding.editTextEmailLogin.text.toString().trim()
            val password = binding.editTextPasswordLogin.text.toString()
            val rememberMe = binding.checkboxRememberMe.isChecked

            if (validateInput(email, password)) {
                Log.d(TAG, "Input validated. Calling ViewModel loginUser.")
                if (::sessionManager.isInitialized) {
                    sessionManager.setRememberMe(rememberMe)
                    if (rememberMe) {
                        sessionManager.saveEmail(email)
                        Log.d(TAG, "Saved email for remember me: $email")
                    } else {
                        sessionManager.clearEmail()
                        Log.d(TAG, "Cleared email for remember me.")
                    }
                } else {
                    Log.e(TAG, "SessionManager not initialized in login button click. Cannot save preferences.")
                }
                authViewModel.loginUser(email, password)
            } else {
                Log.d(TAG, "Input validation FAILED.")
            }
        }

        binding.textViewForgotPassword.setOnClickListener {
            Log.d(TAG, "Forgot Password link clicked")
            if (isAdded) {
                try {
                    // TODO: Thay thế bằng action ID đến màn hình Quên mật khẩu của bạn
                    findNavController().navigate(R.id.action_loginFragment_to_forgotPasswordFragment)
                } catch (e: Exception) {
                    Log.e(TAG, "Navigation to ForgotPassword FAILED", e)
                }
            }
        }

        binding.textViewRegisterLink.setOnClickListener {
            Log.d(TAG, "Register link clicked")
            if (isAdded) {
                try {
                    findNavController().navigate(R.id.action_loginFragment_to_registerFragment)
                } catch (e: Exception) {
                    Log.e(TAG, "Navigation to Register FAILED", e)
                }
            }
        }

        binding.buttonGoogleLogin.setOnClickListener {
            Log.d(TAG, "Google Login button clicked")
            // TODO: Triển khai logic đăng nhập bằng Google
            Toast.makeText(requireContext(), "Chức năng đăng nhập bằng Google sắp ra mắt!", Toast.LENGTH_SHORT).show()
        }
    }

    private fun validateInput(email: String, pass: String): Boolean {
        var isValid = true
        // Validate Email
        if (email.isEmpty()) {
            binding.tilEmailLogin.error = getString(R.string.error_email_trong)
            isValid = false
        } else if (!Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
            binding.tilEmailLogin.error = getString(R.string.error_email_khong_hop_le)
            isValid = false
        } else {
            binding.tilEmailLogin.error = null
        }

        // Validate Password
        if (pass.isEmpty()) {
            binding.tilPasswordLogin.error = getString(R.string.error_mat_khau_trong)
            isValid = false
        } else {
            binding.tilPasswordLogin.error = null
        }
        return isValid
    }

    private fun setupObservers() {
        Log.d(TAG, "setupObservers called")
        authViewModel.authState.observe(viewLifecycleOwner) { state ->
            Log.d(TAG, "AuthState observed: $state")
            binding.progressBarLogin.isVisible = state is AuthState.Loading
            binding.buttonLogin.isEnabled = state !is AuthState.Loading

            when (state) {
                is AuthState.Success -> {
                    Log.d(TAG, "AuthState.Success: ${state.message}")
                    Toast.makeText(requireContext(), state.message ?: "Đăng nhập thành công!", Toast.LENGTH_LONG).show()
                    if (isAdded) {
                        try {
                            Log.d(TAG, "Navigating to Home/Main screen.")
                            // TODO: THAY THẾ R.id.action_loginFragment_to_homeFragment BẰNG ACTION ĐÚNG CỦA BẠN
                            findNavController().navigate(R.id.action_loginFragment_to_homeFragment) // Ví dụ
                        } catch (e: Exception) {
                            Log.e(TAG, "Navigation to Home FAILED", e)
                            Toast.makeText(requireContext(), "Lỗi điều hướng sau đăng nhập.", Toast.LENGTH_SHORT).show()
                        }
                    }
                    authViewModel.resetAuthStateToIdle()
                }
                is AuthState.Error -> {
                    Log.e(TAG, "AuthState.Error: ${state.message}")
                    Toast.makeText(requireContext(), "Lỗi: ${state.message}", Toast.LENGTH_LONG).show()
                }
                AuthState.Idle -> {
                    Log.d(TAG, "AuthState.Idle")
                }
                AuthState.Loading -> {
                    Log.d(TAG, "AuthState.Loading")
                }
            }
        }

        authViewModel.logoutState.observe(viewLifecycleOwner) {isLoggedOut ->
            if(isLoggedOut) {
                Log.d(TAG, "Logout state observed as true.")
                // Xử lý nếu cần, ví dụ xóa thông tin remember me nếu logout từ nơi khác và quay lại đây
                if (::sessionManager.isInitialized) {
                    // sessionManager.setRememberMe(false) // Tùy chọn
                    // sessionManager.clearEmail() // Tùy chọn
                }
                authViewModel.onLogoutCompleted() // Reset lại cờ này trong ViewModel
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        Log.d(TAG, "onDestroyView called")
        if (authViewModel.authState.value is AuthState.Success || authViewModel.authState.value is AuthState.Error) {
            authViewModel.resetAuthStateToIdle()
        }
        _binding = null
    }
}